encoder.go

```diff
@@ -0,0 +1,414 @@
+package rdb
+
+import (
+	"fmt"
+	"math"
+	"sync/atomic"
+	"time"
+)
+
+type Encoder struct {
+	writer       RDBWriter
+	countPos     int64
+	count        int64
+	countWithExp int64
+	backlenBuf   []byte
+	version      string
+	begin        atomic.Bool
+}
+
+func NewFileEncoder(path string, version string) (*Encoder, error) {
+	w, err := newFileWriter(path)
+	if version == "" {
+		return nil, fmt.Errorf("missing Redis version")
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &Encoder{
+		version:    version,
+		writer:     w,
+		backlenBuf: make([]byte, 5),
+		begin:      atomic.Bool{},
+	}, nil
+}
+
+func (s *Encoder) Begin() error {
+	if _, err := s.writer.Write([]byte("REDIS")); err != nil {
+		return err
+	}
+	if _, err := s.writer.Write([]byte(fmt.Sprintf("%04d", Version))); err != nil {
+		return err
+	}
+	if err := s.writeAuxField("redis-ver", s.version); err != nil {
+		return err
+	}
+	if err := s.writeAuxField("redis-bits", "64"); err != nil {
+		return err
+	}
+	if err := s.writeAuxField("ctime", fmt.Sprintf("%d", time.Now().Unix())); err != nil {
+		return err
+	}
+	if err := s.selectDB(0); err != nil {
+		return err
+	}
+	resizeDbPos, err := s.writer.Pos()
```

Comment:
it probably does not matter for us, but should we write module aux info regarding the json module(i see module aux before and after the data in rdb in redis code, but not sure what aux info json module have)? have you tested exporting json data to rdb and importing it in redis?

Reply:
We are already writing JSON module information while creating JSON keys. I successfully exported the JSON values with Upstash Redis, and open-source Redis loaded them without any issues.